File: assign1/scala/src/NamesComputer.scala
import scala.collection.{Map, mutable}

object NamesComputer {

  def averageLengthFunctional(names: List[String]) = {
    if (names.isEmpty) 0 else names.map(_.length).sum.toFloat / names.length
  }

  def namesStartingWithEachLetterFunctional(names: List[String]): Map[String,Int] = {
    names.groupBy(_.substring(0,1)).map {case (nameInitial, nameList) => nameInitial -> nameList.size}
  }

  def averageLengthImperative(names: List[String]) = {
    if (names.isEmpty)
      0
    else { 
      var total = 0.0
      for (name <- names)
        total += name.length
      total.toFloat / names.length
    }
  }

  def namesStartingWithEachLetterImperative(names: List[String]) = {
    val map = new mutable.HashMap[String, Int]()
    for (name <- names)
      map(name.substring(0,1)) = map.getOrElse(name.substring(0,1), 0) + 1
      map
  }
}

File: assign1/ruby/src/names_computer.rb
class NamesComputer

    def average_length_functional(names)
        names.empty? ? 0 : names.sum(&:length) / names.size.to_f
    end
  
    def names_starting_with_each_letter_functional(names)
        names.empty? ?
            Hash.new :
            names.group_by{|name| name.chr}.map{|initial, namelist| [initial,namelist.count]}.to_h
    end

    def average_length_imperative(names)
        return 0.0 if names.empty?

        total = 0.0
        for name in names
            total += name.length
        end

        return total / names.length
    end
      
    def names_starting_with_each_letter_imperative(names)
        return Hash.new if names.empty?

        map = Hash.new
        for name in names
            map[name.chr] = (map[name.chr] || 0) + 1
        end

        return map
    end
end